### PR TITLE
Add command to generate / manage sub-experiments from root experiment config.

### PR DESCRIPTION
**Pull request by HarshTrivedi**
This PR adds a command to generate and easily manage sub-experiments config files from a root experiment config file.

A typical scenarios that (I guess) everyone faces is: I have my experiment model up and running. Nice! But there are lots of things that can be yet configured. I have setting K1 that can take V1a, V1b and V1c values and setting K2 that can take V2a, V2b values and I want to try the combinations of these changes and analyse the results.

One can potentially use `--override` and write a script to try and report these things, but its quite an additional overhead to what this command provides. `--override` is in fact handy for quick checks but not really good when one has to manage bunch of experiment changes and analyse their combinations.

For such scenario one (at least I) typically wants way to:

1. **Part A** Easily generate valid experiment configurations for all 3X2=6 possibilities and organize them in easily identifiable config files.
2. **Part B** Train models using these generated experiment config files.
3. **Part C** And evaluate the trained models on set of test datas and give a analysable/searchable report (preferrably nice html tables).

I have been using Part A, B, C in my work but for now only part A (`generate-subexperiments` command) is PR ready.

To use `generate-subexperiments` command, one needs to pass:

```
1. root_experiment_config   config on which all changed defined in generator_config will be exectued.
2. generator_config         config defining how to generate subexperiments from root experiment.
3. output_directory         where generated config files will be stored.
```

The changes and combinations can be quite complicated at times, but using directly what jsonnet provides makes configuring generator pretty flexible. Let's say my model is `venue-classifier` from [allennlp-as-library](https://github.com/allenai/allennlp-as-a-library-example/blob/master/experiments/venue_classifier.json)  and I want to try out following things:

1. Group 1 (different sized glove embeddings) : 100d, 200d and 300d
2. Group 2 (lstm): bidirectional, unidirectional

So I want to generate 3 X 2 = 6 experiment config files and have them easily organized. There are some caveats here: if lstm `bidirectional` is ON, I would need to double the `input_dim` of `classifier_feedforward`. So in each of these 6 experiment congis, I need to make some post combination corrections. Assume projection is ON and so I don't need to adjust `input_dim` of lstm based on used `embedding_dim`.

The following generator config file would do the required:
```python
{
"combine_changes":[
        { # Group 1 of changes  (3 changes)
                "key_tuple": ["model.pretrained_file", "model.embedding_dim"],
                "value_tuples": [
                        ["'files/glove.6B.100d.txt.gz'", "100"],
                        ["'files/glove.6B.200d.txt.gz'", "200"],
                        ["'files/glove.6B.300d.txt.gz'", "300"]
                ],
                "value_names": ["100d", "200d", "300d"],
                "key_name": "glove_dim"
        },
        { # Group 2 of changes (2 changes)
                "key_tuple": ["model.title_encoder.bidirectional", "model.abstract_encoder.bidirectional"],
                "value_tuples": [
                        ["true", "true"],
                        ["false", "false"]
                ], # value_names can be inferred as well.
                "key_name": "bidirectional"
        }
],
# Sequence of changes for post combination corrections
"post_combine_changes": [
        { # Adjust classifier input after generating each combination.
                "key_tuple": ["model.classifier_feedforward.input_dim"],
                "value_tuple": [
                        """if current.model.title_encoder.bidirectional
                        then current.model.title_encoder.hidden_size * 4
                        else current.model.title_encoder.hidden_size * 2
                        """],
                # following names won't be added in experiment id. No identifier
                # needed because it's a change that all combinations will have.
                "value_name": "adjust",
                "key_name": "ff_input_dim"
        }
]
}
```

This will generate 6 valid experiment configs from root (`venue_classifier.json`) and will identify them with following names:

```
1. -> glove_dim=100d,bidirectional=true
2. -> glove_dim=100d,bidirectional=false
3. -> glove_dim=200d,bidirectional=true
4. -> glove_dim=200d,bidirectional=false
5. -> glove_dim=300d,bidirectional=true
6. -> glove_dim=300d,bidirectional=false
```

There are three parts of a generator config: `"pre_combine_changes"`, `"combine_changes"` and `"post_combine_changes"`. Combinations of groups of changes in `combine_changes` will be used to identify sub experiments. For each of these combinations following changes will be applied in order: sequence of `pre_combine_changes`, combination sequence for current experiment and sequence of `post_combine_changes`.

Because of what jsonnet allows one to do, generation config is very flexible. Each value in a `value_tuple` is a string of valid jsonnet expression. A jsonnet expression need not result in json, but can be string/int/bool etc as well. This makes configuring changes easy. `current` is a special jsonnet variable availble to refer to current state of config just before executing that specific change. Key names and value names make the final generated file names easy for human parsing and managing.

I guess the above example should be self-explanatory. Please check tests for more examples, if not.

### CODE REVIEW

**Comment by matt-gardner** [comment]
Thanks for doing this @HarshTrivedi, this looks like some really useful functionality.  We typically use a separate system for doing this kind of thing internally at AI2, so it's not something we were going to build out ourselves.  And it looks like it'll be a long time before that system is released externally, so it would be nice to have something to handle these issues for AllenNLP users - several people asked me about this at ACL.

Does anyone have any strong opinions on the way this was done?  I don't, so I figure someone with more familiarity with jsonnet, or who has stronger ideas about setting up multiple configs like this, should take a look instead of me.

**Comment by HarshTrivedi** [comment]
@matt-gardner  Thanks, You mentioned earlier about beaker internally here (#1476). I am assuming that it takes care of scheduling / running / evaluating experiments and generating nice analysable reports. I believe that in simple case, this roughly ties together with **Part B** and **Part C** as I mentioned in above comment. But that's not what this PR proposes.

I doubt if a general purpose separate tool with those use cases can generate sub experiments like this command. Generating valid experiments configs (for allennlp) is very much tied together with specifics of allennlp (eg. valid jsonnet) and specifics of your own model / experiment (eg. input of classifier layer should be doubled if I use bidirectional lstm and such idiosyncrasies).

One can surely have a script (specific to his/her experiment) to generate different `--overrides` and have different shell commands which can scheduled/run/evaluated and analysed by whatever tool. But `--overrides` can become very cumbersome even for simple example that I give above. Having different sub experiment config files is much more managable option than managing different `--overrides`.  Alternatively, one can manually generate sub-experiment configs files manually - but is error prone when one wants to explore more than few changes or combinations. 

So independent of what tool one uses for scheduling / running / analysing etc, an easy to use tool for generating valid sub-experiments configs for allennlp (based on what a researcher wants to explore) would be a helpful add-on.

**Comment by joelgrus** [comment]
I'll take a look at it (not today though 😇)

**Comment by joelgrus** [comment]
so I spent some time looking at this today, a few pieces of feedback / for discussion

(1) this feels to me like something that should be in `scripts/` rather than a command.
(2) this feels very complicated to me. like, "I read the code multiple times and I'm still not 100% sure what it's doing or how to use it" complicated. as this PR stands, I don't think anyone would actually be able to use this functionality unless you showed them how. it's possible that all it needs is a tutorial to go with it, but it's also possible that it's more complicated than it needs to be.
(3) I think this will mess up params that are specified by environment variables (that is, by `std.extVar`) -- when you do `Params.from_file` it will populate them based on the machine you're currently running on (or crash if they're not defined there), and then they're gone. If you were counting on them to be on another machine (say, if you run each experiment on some GPU machine), then they're gone. This feels like a problem.
(4) I'm not a huge fan of `ast.literal_eval`. I don't really understand the part of the code that uses it, and I feel like that makes it even harder to understand.

--

along the "too complicated" lines, isn't a lot of this available as built-in jsonnet functionality? that is, couldn't you just bake all of this into your jsonnet like

```
    "bidirectional": std.extVar("BIDIRECTIONAL"),
    ....
    ....
    "input_dim": if $.model.title_encoder.bidirectional then $.model.title_encoder.hidden_size * 4 else $.model.title_encoder.hidden_size * 2,
```

and then just iterate over combinations of environment variables? wouldn't that accomplish the same thing in a much simpler way?

**Comment by joelgrus** [comment]
that long-winded comment was me thinking aloud as I read the PR, but really what it boils down to is: what problem does this solve that jsonnet doesn't already solve?

**Comment by HarshTrivedi** [comment]
> 1) this feels to me like something that should be in scripts/ rather than a command. 

Sure, I can change that later.

> (2) this feels very complicated to me. like, "I read the code multiple times and I'm still not 100% sure what it's doing or how to use it" complicated. as this PR stands, I don't think anyone would actually be able to use this functionality unless you showed them how. it's possible that all it needs is a tutorial to go with it, but it's also possible that it's more complicated than it needs to be.

In terms of usage, it's actually very simple. I believe the reason it seems complicated to use is that I tried to show complicated examples first that leverage jsonnet features. Let me try again by starting with simple motivating examples iteratively showing why and in what scenario particular feature would be required. I will get back with examples tomorrow...

> (3) I think this will mess up params that are specified by environment variables (that is, by std.extVar) -- when you do Params.from_file it will populate them based on the machine you're currently running on (or crash if they're not defined there), and then they're gone. If you were counting on them to be on another machine (say, if you run each experiment on some GPU machine), then they're gone. This feels like a problem.

If I understand you correctly, that shouldn't be a problem.  `generate-subexperiments` generates fully specified sub-experiment config files. The output experiment config files would not have such expressions `std.extVar("BIDIRECTIONAL")`- they would be plain jsons with everything specified based on generator config and root config. And if something like `std.extVar("BIDIRECTIONAL")` is actually used in generator_config_file / root experiment file it would work as expected.

> (4) I'm not a huge fan of ast.literal_eval. I don't really understand the part of the code that uses it, and I feel like that makes it even harder to understand.

Sure, I knew you would point `literal_eval` out : ) .. but I don't see any alternative. `execute` method takes in a source experiment config and executes a change on it which that Change-instance encodes. For simplicity consider Change as a `key` and `value` pair instead of `key_tuple` and `value_tuple`. `key` refers to key that you want to change: eg 'model.seq2vec_encoder.bidrectional', and value is the hard-coded or computed value that you want to set. The value can be any valid string: eg. "true", "false", ("[1, 2, 3]" if arrays were permitted!) or jsonnet expression "! true", "! false" or something that requires complicated referrence: "if current.something then true else false".

The following line: `value_evaluated = _jsonnet.evaluate_snippet("", define_current_var+value).strip()` returns a evaluation of jsonnet expression. Eg. `value = '{"hello": 1+2}'` would yeild `value_evaluated = '{"hello": 3}'`. We generally parse it by `json.loads` since we expect output of it to be dict. However, jsonnet expression can lead to any thing as we want: boolean, string, list etc hence `json.loads` is not appropriate.  Eg. `value = '!false'` would lead to `value_evaluated = 'true'`. `value_evaluated` is actually is string of python object which we want to set but it has json based constants since it came out of jsonnet. Hence we convert: "true"->"True", "null"->"None" etc and finally use `ast.literal_eval` to parse it as string/dict/list/boo etc. 

> (5) couldn't you just bake all of this into your jsonnet like .. and then just iterate over combinations of environment variables? wouldn't that accomplish the same thing in a much simpler way?

In terms of usage, IMO it would be much more cumbersome to manage, and it doesn't support many things that script would support.

1. Firstly, to use something like `std.extVar("BIDIRECTIONAL")` you would need to set and manage this environment variable explicitly. But for `currrent.model.seq2vec.bidirectional` you don't need anything. The referencing says that see that in current  state of experiment config, bidirectional is true or not and decide based on that. By adding `current.` variable to jsonnet you are saying that you can reference any part of the current experiment config arbitrarily with full flexibility.
2. Secondly, it's not just about iterating over combinations - the script gives you easily managable handle on giving names to the changes and organizing them in separate files. One can actually open these files, look at their names and check if things are correct. Changes identified by simplified names make things much more easy to manage.

3. It also gives you way to group the logically groupable changes so that they don't get combinated and one can identify them easily. .. I guess examples I give later would make this clear.

4. ..

Let me get back tomorrow with separated and motivating examples for this kind of script and perhaps a short tutorial ...

**Comment by HarshTrivedi** [comment]
> that long-winded comment was me thinking aloud as I read the PR, but really what it boils down to is: what problem does this solve that jsonnet doesn't already solve?

Sure, I get you. ...  jsonnet is not mandatory. Infact, I used this script for long time w/o jsonnet  for many things. So purpose of jsonnet and this scripts are quite different - jsonnet features are like addon which make it better.  ... Let me make some motivating examples and get back : )

**Comment by HarshTrivedi** [comment]
This is quite verbose .. but it is more in the lines of a tutorial as you suggested ...

# What's the use?

**To put things in context:** A typical scenario that one usually faces is - I have my experiment config file working up and running but there are yet many things that can be configured for which there is no obvious choice. Only aparant way is to see empirically what works better and obtain the analysable trends. This PR is first step towards easy explorations of sub-experiments

A typical work flow for such a requirement would be:

1. Generate subexperiment config files and organize them with easily identifiable file names. (`generate-subexperiments`)
2. Train different models for each of the generated sub-experiment config files (`train-subexperiments`)
3. Evaluate these models on one or more evaluation datasets. (`evaluate-subexperiments`)
4. Generate analysable html reports that clearly say what was the change from root-experiment and what was the effect. (`analyse-subexperiments`)

The `generate-subexperiments` commands deals with part (1) of this requirement. It does so by taking in two files: `root_experiment_config` and `generator_config` and outputs all the identifiable sub-experiment config files in some specified `subexperiments_directory`.

The commands for (2), (3) and (4) are not in PR yet. But they basically, consume `subexperiments_directory` and do their required task.

The `generator_config` file gives a **structured and flexible** way to specify what your sub-epxeriments exploration will be. Following section explains it.

# Usage of generator config

For most examples, I will be using venue_classifier.json` from [here](https://github.com/allenai/allennlp-as-a-library-example/blob/master/experiments/venue_classifier.json) as root experiment config.

## Change

The underlying idea of generating **a** sub-experiment from a root experiment is a sequence of `Change`s. These `Change`s will be executed starting from root experiment **sequentially** to finally give use **one** sub-experiment. Something like: (`root_experiment_config` -> `Change1`, `Change2` ... `ChangeN` -> `sub_experiment_config`). So a `sub_experiment` config is nothing but a sequence of changes executed on `root_experiment_config`.

A `Change` simply put encodes which key in current config should be set to which value on execution of this change. It consists of `key_tuple`, `value_tuple`, `key_name` and `value_name`.  `key_tuple` is a **tuple** of keys and `value_tuple` is a **tuple** of values in which individual `key`, `value`'s have following roles.

- A key (str) refers to what "key" of `source_config` one wants to change.
    - Eg. `"model.title_encoder.bidirectional"`, `"trainer.batch_size"` etc.
- A value (str) refers to what "value" one wants to set that key to.
    - Eg. `"true"`, `"10"`, `"[20, 20, 20]"`, `"['relu']"`, `"'path/to/file'"` etc. A value is always a string irrespective of you want the value to be int / list/ dict/ str. See difference between `"true"` and `"'path/to/file'"` (extra `'` in `"'path/to/file'"`). Reason for this will be clear later.
- `key_name` and `value_name` give easy way of identifying the `Change`. Since a supexperiment is a unique sequence of changes, names of constituent changes of subexperiment give easily identifiable name to that sub-experiment.

Here, `key_tuple` and `value_tuple` are of same cardinality and ith key refers to ith value.

### Change (from params): Example 1

Lets say we want to construct `change` for turning Off bidirectionality of seq2vec-encoder for `title_encoder` :
```python
# Change: Make title_encoder unidirectional
{
    "key_tuple": ["model.title_encoder.bidirectional"],
    "value_tuple": ["false"],
    "key_name": "title-enc-bidirectional"
    "value_name": "false",
}
```

When this `change` is `executed` on our `venue_classifier` config, it will return a config file with unidirectional title encoder and all else unchanged. The name of this change will be: `title-enc-bidirectional=false`, which is representative of what it does on execution.

### Change (from params): Example 2

But, actually we have 2 seq2vec encoders in our model. Let's say we want to construct a change for turning bidirectional Off for both of them and identify this change as `all_bidirectional=false`.

```python
# Change: Make title_encoder and abstract_encoder unidirectional
{
    "key_tuple": ["model.title_encoder.bidirectional", "model.abstract_encoder.bidirectional"],
    "value_tuple": ["false", "false"],
    "key_name": "all_bidirectional"
    "value_name": "false",
}
```
See that corresponding entries in key_tuple and value_tuple form the corresponding pairs.
The name of this change will be: `all_bidirectional=false`.

Side note: This abstraction is implemented in `Change` class.

## Sequence of Changes

A sequence of changes is just list (sequence) of changes. For example, we want to make a sequence of changes for (1) make all seq2vec encoders unidirectional and (2) make embeddings trainable, it would be something as following:

```python
# Sequence of following changes:
# (1) making all seq2vec encoders unidirectional
# (2) make embedding trainable
[
{
    "key_tuple": ["model.title_encoder.bidirectional", "model.abstract_encoder.bidirectional"],
    "value_tuple": ["false", "false"],
    "key_name": "all_bidirectional"
    "value_name": "false",
}
{
    "key_tuple": ["model.text_field_embedder.tokens.trainable"],
    "value_tuple": ["true"],
    "key_name": "embeddings_trainable"
    "value_name": "true",
}
]
```

Name of sequence of changes reflects what changes it constitutes and in what order. The name of this sequence of changes will be: `all_bidirectional=false,embeddings_trainable=true`. As one can expect from name: when it will be `execute`d on `root_experiment_config`, it will make all seq2vec encoders unidirectional and make embeddings trainable.

Side note: This abstraction is implemented in `SequenceOfChanges` class.

## Group of Changes

A group of changes are the changes that one wants to take effect one at at time. Eg. lets say I want to try three types of `title_encoder`'s: "lstm", "gru", "rnn", then the `key` is `"model.title_encoder.type"` and values to be explored are `"'lstm'"`, `"'gru'"`, `"'rnn'"`. Clearly, a group will have 1 `key_tuple` and many `value_tuple`s that one wants to explore - one at a time. For this example, following params would construct such a group:

```python
{
    "key_tuple": ["model.title_encoder.type"],
    "value_tuples": [
                 ["lstm"], # value_tuple 1
                 ["gru"],  # value_tuple 2
                 ["rnn"],  # value_tuple 3
              ]
    "key_name": "title_encoder_type"
    "value_names": ["lstm", "gru", "rnn"],
}
```

Side note: This abstraction is implemented in `GroupOfChanges` class.

## Putting things together

Since we know what `Change`, `SequenceOfChanges` and `GroupOfChanges` are, building generator-config file is straight forward. A generator-config file has 3 parts. Following are the type annotations:

```python
{
   "pre_combine_changes": [], # Sequence of Changes
   "combine_changes": []  # List of Group of Changes
   "post_combine_changes": [] # Sequence of Changes
}
```

`pre_combine_changes` and `post_combine_changes` are pre and post combine hooks for executing sequence of changes before and after making combinations. For now, let's skip them - they are not useful for now.

The list of group/s in `"combine_changes"` is central for generating sub-experiments. The generated subexperiments will be combinations from these groups.

### Example 1

Lets say we want to try the combinations of:

1. try *lstm* and *gru* seq2vec encoders
2. try *bidirectional* and *unidirectional* seq2vec encoders


First step is to generate `GroupOfChanges` for (1) and (2).

A group of changes for (1) will be as follows:

```
    # Group 1: X 2 choices
    { # try *lstm* and *gru* seq2vec encoders
    "key_tuple": ["model.title_encoder.type", "model.abstract_encoder.type"],
    "value_tuples": [
                 ["lstm", "lstm"], # value_tuple 1
                 ["gru", "gru"  ], # value_tuple 2
              ]
    "key_name": "encoders_type"
    "value_names": ["lstm", "gru"],
    }
```

And group of changes for (2) will be as follows:

```
    # Group 2: X 2 choices
    { # try *bidirectional* and *unidirectional* seq2vec encoders
    "key_tuple": ["model.title_encoder.bidirectional", "model.abstract_encoder.bidirectional"],
    "value_tuples": [
                 ["true", "true"],     # value_tuple 1
                 ["false", "false"  ], # value_tuple 2
              ]
    "key_name": "all_bidirectional"
    "value_names": ["true", "false"],
    }
```

Finally, we just push these two in `combine_changes` list since we want to make combinations of these groups. Following is the full generator config file for such a generation:

```python
"combine_changes": [
    # Group 1: X 2 choices
    { # try *lstm* and *gru* seq2vec encoders
    "key_tuple": ["model.title_encoder.type", "model.abstract_encoder.type"],
    "value_tuples": [
                 ["lstm", "lstm"], # value_tuple 1
                 ["gru", "gru"  ], # value_tuple 2
              ]
    "key_name": "encoders_type"
    "value_names": ["lstm", "gru"],
    }

    # Group 2: X 2 choices
    { # try *bidirectional* and *unidirectional* seq2vec encoders
    "key_tuple": ["model.title_encoder.bidirectional", "model.abstract_encoder.bidirectional"],
    "value_tuples": [
                 ["true", "true"],     # value_tuple 1
                 ["false", "false"  ], # value_tuple 2
              ]
    "key_name": "all_bidirectional"
    "value_names": ["true", "false"],
    }

]
```

**Pictorially**, it is something like following:

```
                   encoders_type=lstm     all_bidirectional=true
root_experiment -> encoders_type=gru  ->  all_bidirectional=false  -> sub_experiment

                      [ Group 1]               [ Group 2]
                   (choose 1 here)           (choose 1 here)
```

This will generate 4 sub-experiments with following representative names (and as expected behaviour). These names will be present in the file names which these configs will be dumped to.

```
encoders_type=lstm,all_bidirectional=true
encoders_type=lstm,all_bidirectional=false
encoders_type=gru,all_bidirectional=true
encoders_type=gru,all_bidirectional=false
```

More concretely, the following will be the dumped file names:

```
<root_experiment_name>.encoders_type=lstm,all_bidirectional=true.json
<root_experiment_name>.encoders_type=lstm,all_bidirectional=false.json
<root_experiment_name>.encoders_type=gru,all_bidirectional=true.json
<root_experiment_name>.encoders_type=gru,all_bidirectional=false.json
```

This actually makes it very easy to debug, manage and analyse inidividual subexperiments. This is because, now we have an easy handle on individual sub-experiment config files with file-names representative of what is being explored in that particular sub-experiment.

Note that I gave here short example that results in only 4 subexperiments for brevity. Actualy one might want to explore much more (I typically try to analyse 10-15 subexperiments in one time). It's easy to imagine that w/o a structured way of organizing as I show above, sub-experiments exploration can become quite cumbersome.


### Example 2

Using combinations isn't necessary to be used. In some cases, one could just want to try different values. Let's say one wants on only explore all types of sec2vec encoder types and that it.

The following generator config file will do the required:

```python
"combine_changes": [
    # Group 1: X 2 choices
    { # try *lstm*, *gru*, *rnn* seq2vec encoders
    "key_tuple": ["model.title_encoder.type", "model.abstract_encoder.type"],
    "value_tuples": [
                 ["lstm", "lstm", "lstm"], # value_tuple 1
                 ["gru", "gru", "gru"],    # value_tuple 2
                 ["rnn", "rnn", "rnn"],    # value_tuple 3
              ]
    "key_name": "encoders_type"
    "value_names": ["lstm", "gru", "rnn"],
    }
]
```

**Pictorially**, it is something like following:

```
                   encoders_type=lstm
root_experiment -> encoders_type=gru  -> sub_experiment
                   encoders_type=rnn

                      [ Group 1]
                   (choose 1 here)
```


This will generate 3 sub-experiments with following names (and as expected behaviour)
```
encoders_type=lstm
encoders_type=gru
encoders_type=rnn
```

Using this one can simply create different types of sub-experiments with or without combinations. `post_combine_changes` becomes useful with usage of jsonnet as shown in next section. Please note that using jsonnet features is not always required. Extrapolating the above 2 examples, it should be evident that even w/o it, one can do fairly good amont of sub-experiment config files generation.

## Using Jsonnet Features with it

### Without Reference:

The `value` in `Change` is actually any valid jsonnet expression in string. jsonnet expression do not need to yield a json / dict always. It can yield any type of data: string / int / bool / list etc. This makes using or not using jsonnet seamless.

Lets say we know some `model.feedforward` has 3 `num_layers` and we want to set dropout of 0.2 for all the layers. The following `Change` would work.

```python
# Change: Set dropout
{
    "key_tuple": ["model.feedforward"],
    "value_tuple": ["[0.2  for _ in std.range(1, 3)]"],
    "key_name": "dropout"
    "value_name": "0.2",
}
```
Note: value `"[0.2  for _ in std.range(1, 3)]"` is a valid jsonnet expression (try [here](https://jsonnet.org/)) and hence it works.


### With Reference:

In above example, we assumed we knew that `model.feedforward` has 3 `num_layer`s. However, it's possible that it could have been changed before. In such cases, one can use `current` jsonnet variable that refers to the current state of experiment config. For example if a subexperiment is as follows: `root_experiment` -> `Change1` -> `Change2` -> `Change3` -> `sub_experiment` and we use current variable in `Change2`, it refers to the state of experiment config right after `Change1` is executed and before `Change2` is exectued.

For example, if unlike above example, we don't know the num_layers of feedforward upfront / or it's dynamic being changed in previous changes, then following can be used. It assures that dropout setting is consistent with `num_layers` independent of what `num_layers` were set to in previous cases.

```python
# Change: Set dropout
{
    "key_tuple": ["model.feedforward"],
    "value_tuple": ["[0.2  for _ in std.range(1, current.feedforward.num_layers)]"],
    "key_name": "dropout"
    "value_name": "0.2",
}
```

Below is a more concrete example would explain the need of `current` variable.

Lets say I want to explore the following:

1. try model.feedforward.num_layers to 2, 3, 4  [group of 3 changes]
2. try dropouts 0.0, 0.2  [group of 2 changes]

This combination will have 3 X 2 = 6 sub-experiments. Note that value of second group actually depends on what was being set in the first group. `num_layers` are dynamic and can't be known upfront. Leveraging `current.` variable, the following `generator config` solves this problem.

```python
"combine_changes": [
    # Group 1: X 3 choices
    { # model.feedforward.num_layers to 2, 3, 4
    "key_tuple": ["model.feedforward.num_layers"],
    "value_tuples": [
                 ["2"], # value_tuple 1
                 ["3"], # value_tuple 2
                 ["4"], # value_tuple 3
              ]
    "key_name": "ff_num_layers"
    "value_names": ["1", "2", "3"],
    }

    # Group 2: X 2 choices
    { # try dropouts 0.0, 0.2 for feedforward
    "key_tuple": ["model.feedforward.dropouts"],
    "value_tuples": [
                 ["[0.0  for _ in std.range(1, current.feedforward.num_layers)]"],  # value_tuple 1
                 ["[0.2  for _ in std.range(1, current.feedforward.num_layers)]"],  # value_tuple 2
              ]
    "key_name": "all_bidirectional"
    "value_names": ["true", "false"],
    }
]
```

The scenario is something like following:

```
                   ff_num_layers=2     dropouts=0.0
root_experiment -> ff_num_layers=3 ->  dropouts=0.2  -> sub_experiment
                   ff_num_layers=4

                      [ Group 1]        [ Group 2]
                   (choose 1 here)     (choose 1 here)

                                  ^
                                  |

                          `current` var in group 2
                          refers  to  config after
                          choice from group 1  has
                          been made  and  executed

```

Note, the usage of `current.` variable is essential in this case because in that specific sub-experiment `num_layers` cannot be known upfront. The variable `current.` refers to state of experiment config just before the execution of that change in sequence of changes. This allows to inspect any field / key / etc in current state of config and make decisions based on that.

Such a scenario also calls for usage of `post_combine_changes`. Note that even though, we adjust `dropout` according to num_layers set in that sub-experiment, the final config is not yet valid. One needs to make corrections / adjustments to the `activations` and `hidden_sizes` based on what `num_layers` were set to. However, it is just mandatory post correction to make sub-experiment valid, and not something that we are deliberately try to explore. All sub-experiments will need such a post correction, so we add this sequnce in `post_combine_changes` hook. In entirity the generator config file would look like:


```python
{

  # combine_changes is same as above.
  "combine_changes": [
      # Group 1: X 3 choices
      { # model.feedforward.num_layers to 2, 3, 4
      "key_tuple": ["model.feedforward.num_layers"],
      "value_tuples": [
                   ["2"], # value_tuple 1
                   ["3"], # value_tuple 2
                   ["4"], # value_tuple 3
                ]
      "key_name": "ff_num_layers"
      "value_names": ["1", "2", "3"],
      }

      # Group 2: X 2 choices
      { # try dropouts 0.0, 0.2 for feedforward
      "key_tuple": ["model.feedforward.dropouts"],
      "value_tuples": [
                   ["[0.0  for _ in std.range(1, current.feedforward.num_layers)]"],  # value_tuple 1
                   ["[0.2  for _ in std.range(1, current.feedforward.num_layers)]"],  # value_tuple 2
                ]
      "key_name": "all_bidirectional"
      "value_names": ["true", "false"],
      }
  ]

  # Sequnce of Changes for Post Combine Corrections.
  "post_combine_changes": [
    {
      "key_tuple": ["model.feedforward.activations"],
      "value_tuple": ["['relu'  for _ in std.range(1, current.feedforward.num_layers)]"]
    }
    {
      "key_tuple": ["model.text_field_embedder.tokens.trainable"],
      "value_tuple": ["[200  for _ in std.range(1, current.feedforward.num_layers)]"],
    }
  ]

}
```

This will generate the following sub-experiment names:

```
ff_num_layers=2,all_bidirectional=true
ff_num_layers=2,all_bidirectional=false
ff_num_layers=3,all_bidirectional=true
ff_num_layers=3,all_bidirectional=false
ff_num_layers=4,all_bidirectional=true
ff_num_layers=4,all_bidirectional=false
```

Note that post combination changes won't count towards the sub-experiment name since they are common across all sub-experiments.


### Summary of Generator Config:


To **summarize**, a general generator config builds on idea of sequence of Changes and if all sections are used pictorially it would look like:

```
root_experiment --> [ pre-change-1 -> pre-change-2 .. pre-change-N ]   -->  [  Group 1       Group 2    ..  Group 3    ]      -->  [ post-change-1 -> post-change-2 .. post-change-N ]  -->  sub-experiment
                                                                                 .             .             .
                                                                                 .             .             .
                                   "pre_combine_changes"                      choose-1     choose-1    ..  choose-1                        "post_combine_changes"

                                                                                             "combine_changes"
```


# Importance of representative names

From above examples, I would be evident that names of sub-experiments (composed from constituent changes) actually makes it easily managable. Having viewable sub-experiment config files with relevant names makes it easy to verify that sub-experiment is as expected as file-name suggests. However, that's not only benefit:

Although `report_experiments` command is not PR-ready, the final report makes a lot of sense if you have given names to individual changes that you are exploring.  Eg. If there are 4 subexperiments being explored:
```
all_bidirectional=true,embedding_trainable=true
all_bidirectional=true,embedding_trainable=false
all_bidirectional=false,embedding_trainable=true
all_bidirectional=false,embedding_trainable=false
```
it would generate searchable html report for each configured evauation dataset as (check #1566 for sample example):

| all_bidirectional|embedding_trainable|metric_1|metric_2|
|----------|:-------------:|------:|------:|
| true |  true    | 0.8 | 0.6
| true |  false   | 0.3 | 0.7
| false | true    | 0.6 | 0.3
| false | false   | 0.7 | 0.5

The ability to club changes in one logical name also helps analysis. They abstract out internal details of what that change constitutes. Eg. `all_bidirectional` actually means bidirectional of both abstract and title encoder but it's a redundant detail for analysis and organizing the sub-experiment file names.


# Other Comments

> what problem does this solve that jsonnet doesn't already solve?

I hope it's clear from these examples that the purposes are fundamentally different. jsonnet features are addon to sub-experiment generation command and one can handle fairly good scenarios of sub-epxeriment generation even without jsonnet features. I specifically delayed using jsonnet in examples to clarify this point.

> Why not just bake all of this into your jsonnet ... and iterate over combinations of environment variables?

The 3 reasons I gave in above comment and examples here support why not. Lastly, abstractions that `Change`, `SequenceOfChanges` and `GroupOfChanges` give to generate subexperiments from root experiment appears much cleaner to me and are surely more flexible.

> "I read the code multiple times and I'm still not 100% sure what it's doing or how to use it"

I was mistaken that 1 example in original PR and test cases would be sufficient to explain purpose and syntax! But after following examples in **this** comment, **using** generator-config should seem very straight-forward. For reviewing implementation, looking at `Change`, `SequenceOfChanges` and `GroupOfChanges` classes and comparing with their utility from examples above should suffice.

> I don't think anyone would actually be able to use this functionality unless you showed them how ..

I can make a tutorial out of this comment, if it seems appropriate.

> ...  but it's also possible that it's more complicated than it needs to be.

The examples I give above are still simple enough to make a point of each functionality and explain the short pipeline. For very small / simple examples (**implementation wise**) it might seem that a structured approach to sub-experiment generation is an overkill. But I would request to consider the possibility that one might need to handle 10's of sub-experiments explorations and need a way to handle/manage and analyse them. As far as usage / end-user friendliness is concerned, I hold that usage should be quite straightforward for small / big examples after following this comment.

**Comment by joelgrus** [comment]
thanks for all the explanation, but I still don't get what your code gives me that I can't get using basic jsonnet features. 

here I modified the venue_config.jsonnet to be parameterized:

https://gist.github.com/joelgrus/5ce9af24373a23ba97436874a9c3aec8

and I wrote a very light python script to generate the corresponding configs. (I didn't generate names for them, but that would be one more line of code).

what important features is my version missing?

**Comment by HarshTrivedi** [comment]
One can always write a specific script + parameterized root config json for his/her own specific purpose and can do all the things as I show above.

The point of my script is to not have to write specific parameterized root config or specifc script. But rather have generalized way of configuring what one wants to explore by defining only one config.
This includes: what keys will be set to what value, names and (not included yet) evaluation datasets to explore, analyse and report these generated sub-experiment configs on. All things pertaining to
train / evaluate / analyse / report explorations of the root experiment in **one** place (and **separate** from root config).

An equivalent config for above command would be:

```json
{

  "combine_changes": [
      {
      "key_tuple": ["model.title_encoder.bidirectional", "model.abstract_encoder.bidirectional"],
      "value_tuples": [
                   ["true", "true"],
                   ["false", "false"]
                ],
      }
      {
      "key_tuple": ["model.text_field_embedder.tokens.trainable"],
      "value_tuples": [
                   ["true"],
                   ["false"],
                ],
      }
      {
      "key_tuple": ["model.title_encoder.type"],
      "value_tuples": [
                   ["'gru'"],
                   ["'lstm'"],
                ],
      }
  ]
}
```

and would not require any specific parameterization / variable defining / changes in root config. It is self-sufficient and quite succinct on what one wants to explore starting from a root config.

If I compare your way and mine, main difference is:

- I decide what key will be set to what value in one separate config.
- You define variables in one config (SUB), pass variables in to root config and leave decision of what key to set to what value in root config.

If you make a generally usable script out of your script, it will need to take 2 inputs: SUB like config file and parameterized root config both of which will have to be configured specific to one's requirement.
Although your is simple implementation-wise, I think this PR one is easier to use. If you think otherwise, I would be fine closing this ....

**Comment by HarshTrivedi** [comment]
@joelgrus no problem .. Thank you for the feedback!
